src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,857 @@
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": 1
+  }
+  ```
+
+* Preserved array `$values` property contains object
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": {}
+  }
+  ```
+
+## JSON Objects if not Collection (Class | Struct | Dictionary) - On Deserialize (and Serialize?)
+
+* `$ref` **Valid** under conditions:
+  * must be the only property in the object.
+
+* `$id` **Valid** under conditions:
+  * must be the first property in the object
+
+* `$values` **Not Valid**
+
+* `$.*` **Valid**
+
+Note: For Dictionary keys on serialize, should we allow serializing keys `$id`, `$ref` and `$values`? If we allow it, then there is a potential round-tripping issue.
+Sample of similar issue with `DictionaryKeyPolicy`:
+```cs
+public static void TestDictionary_Collision() 
+{
+    var root = new Dictionary<string, int>();
+    root["helloWorld"] = 100;
+    root["HelloWorld"] = 200;
+
+    var opts = new JsonSerializerOptions
+    {
+        DictionaryKeyPolicy = JsonNamingPolicy.CamelCase
+    };
+
+    string json = JsonSerializer.Serialize(root, opts);
+    Console.WriteLine(json);
+    /* Output:
+    {"helloWorld":100,"helloWorld":200} */
+
+    // Round tripping issue
+    root = JsonSerializer.Deserialize<Dictionary<string, int>>(json);
+}
+```
+
+
+## JSON Object if Collection - On Deserialize
+
+* `$ref` **Valid** under conditions:
+  * must be the only property in the object.
+
+* `$id` **Valid** under conditions:
+  * must be the first property in the object.
+
+* `$values` **Valid** under conditions:
+  * must be after `$id`
+
+* `$.*` **Not Valid**
+
+
+## Immutable types
+Since these types are created with the help of an internal converter, and they are not parsed until the entire block of JSON finishes; nested reference to these types is impossible to identify, unless you re-scan the resulting object, which is too expensive.
+
+With that said, the deserializer will throw when it reads `$id` on any of these types. When serializing (i.e. writing) those types, however, they are going to be preserved as any other collection type (`{ "$id": "1", "$values": [...] }`) since those types can still be parsed into a collection type that is supported.
+
+Note: By the same principle, `Newtonsoft.Json` does not support parsing JSON arrays into immutables as well.
+Note 2: When using immutable types and `ReferenceHandling.Preserve`, you will not be able to generate payloads that are capables of round-tripping.
+
+* **Immutable types**: i.e: `ImmutableList` and `ImmutableDictionary`
+* **System.Array**: i.e: `Array<T>` and `T[]`
```

Comment:
Are collections that don't implement `IList`/`IDictionary`, but don't use internal converters--like `Stack<T>` and `HashSet<T>`--supported for this feature?

Reply:
Yes, all collections, as long as they don't rely on a converter, should be compatible with the feature.